// grammarkdown/out/lib/performance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("./core");
var timestamp = Date.now ? function () { return Date.now(); } : function () { return +(new Date()); };
var counts;
var marks;
var measures;
var performanceStart;
reset();
function mark(markName) {
    core_1.Dictionary.increment(counts, markName);
    core_1.Dictionary.set(marks, markName, timestamp());
}
exports.mark = mark;
function measure(measureName, startMark, endMark) {
    var end = endMark && core_1.Dictionary.get(marks, endMark) || timestamp();
    var start = startMark && core_1.Dictionary.get(marks, startMark) || performanceStart;
    core_1.Dictionary.increment(measures, measureName, end > start ? end - start : start - end);
}
exports.measure = measure;
function getMark(markName) {
    return core_1.Dictionary.get(marks, markName) || 0;
}
exports.getMark = getMark;
function getMarks() {
    return core_1.Dictionary.cloneReadonly(counts);
}
exports.getMarks = getMarks;
function getDuration(measureName) {
    return core_1.Dictionary.get(measures, measureName) || 0;
}
exports.getDuration = getDuration;
function getMeasures() {
    return core_1.Dictionary.cloneReadonly(measures);
}
exports.getMeasures = getMeasures;
function reset() {
    counts = new core_1.Dictionary();
    marks = new core_1.Dictionary();
    measures = new core_1.Dictionary();
    performanceStart = timestamp();
}
exports.reset = reset;
