// prex/out/lib/utils.js
/*! *****************************************************************************
Copyright (c) Microsoft Corporation.
Licensed under the Apache License, Version 2.0.

See LICENSE file in the project root for details.
***************************************************************************** */
"use strict";
function isString(value, optional) {
    return isTypeOf(value, "string", optional);
}
exports.isString = isString;
function isNumber(value, optional) {
    return isTypeOf(value, "number", optional);
}
exports.isNumber = isNumber;
function isBoolean(value, optional) {
    return isTypeOf(value, "boolean", optional);
}
exports.isBoolean = isBoolean;
function isObject(value, optional) {
    return isTypeOf(value, "object", optional);
}
exports.isObject = isObject;
function isFunction(value, optional) {
    return isTypeOf(value, "function", optional);
}
exports.isFunction = isFunction;
function isIterable(value, optional) {
    return isMissing(value) ? !!optional : typeof value === "object" && Symbol.iterator in value;
}
exports.isIterable = isIterable;
function isInstance(value, constructor, optional) {
    return isMissing(value) ? !!optional : value instanceof constructor;
}
exports.isInstance = isInstance;
function isTypeOf(value, typeTag, optional) {
    return isMissing(value) ? !!optional : typeof value === typeTag;
}
function isMissing(value) {
    return value === null || value === undefined;
}
exports.isMissing = isMissing;
