// ecmarkup/lib/NonTerminal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Builder_1 = require("./Builder");
/*@internal*/
class NonTerminal extends Builder_1.default {
    constructor(spec, node, namespace) {
        super(spec, node);
        this.params = node.getAttribute('params');
        this.optional = node.hasAttribute('optional');
        this.namespace = namespace;
    }
    static enter({ spec, node, clauseStack }) {
        const clause = clauseStack[clauseStack.length - 1];
        const namespace = clause ? clause.namespace : spec.namespace;
        const nt = new NonTerminal(spec, node, namespace);
        spec._ntRefs.push(nt);
    }
    build() {
        const name = this.node.textContent;
        const id = 'prod-' + name;
        const entry = this.spec.biblio.byProductionName(name, this.namespace);
        if (entry) {
            this.node.innerHTML = '<a href="' + entry.location + '#' + entry.id + '">' + name + '</a>';
            this.entry = entry;
        }
        else {
            this.node.innerHTML = name;
        }
        let modifiers = '';
        if (this.params) {
            modifiers += '<emu-params>[' + this.params + ']</emu-params>';
        }
        if (this.optional) {
            modifiers += '<emu-opt>opt</emu-opt>';
        }
        if (modifiers.length > 0) {
            const el = this.spec.doc.createElement('emu-mods');
            el.innerHTML = modifiers;
            this.node.appendChild(el);
        }
    }
}
NonTerminal.elements = ['EMU-NT'];
exports.default = NonTerminal;
