// grammarkdown/out/lib/emitter/html.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var emitter_1 = require("./emitter");
var tokens_1 = require("../tokens");
var HtmlEmitter = (function (_super) {
    __extends(HtmlEmitter, _super);
    function HtmlEmitter() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.extension = ".html";
        return _this;
    }
    HtmlEmitter.prototype.emitSourceFile = function (node) {
        this.writer.write("<div class=\"grammar\">");
        this.writer.indent();
        this.writer.writeln();
        _super.prototype.emitSourceFile.call(this, node);
        this.writer.dedent();
        this.writer.writeln();
        this.writer.write("</div>");
    };
    HtmlEmitter.prototype.emitProduction = function (node) {
        var linkId = this.resolver.getProductionLinkId(node.name);
        this.emitLinkAnchor(linkId);
        this.writer.write("<div class=\"production\">");
        this.writer.indent();
        this.writer.writeln();
        this.writer.write("<span class=\"nonterminal\">");
        this.emitIdentifier(node.name);
        this.writer.write("</span>");
        this.emitNode(node.parameterList);
        if (node.colonToken) {
            this.writer.write("<span class=\"punctuation\">");
            switch (node.colonToken.kind) {
                case tokens_1.SyntaxKind.ColonToken:
                    this.writer.write(" :");
                    break;
                case tokens_1.SyntaxKind.ColonColonToken:
                    this.writer.write(" ::");
                    break;
                case tokens_1.SyntaxKind.ColonColonColonToken:
                    this.writer.write(" :::");
                    break;
            }
            this.writer.write("</span>");
        }
        this.writer.writeln();
        this.emitNode(node.body);
        this.writer.dedent();
        this.writer.writeln();
        this.writer.write("</div>");
        this.emitTrailingHtmlTriviaOfNode(node);
        this.writer.writeln();
    };
    HtmlEmitter.prototype.emitParameterList = function (node) {
        this.writer.write("<span class=\"parameter-list\">");
        this.writer.write("[");
        for (var i = 0; i < node.elements.length; ++i) {
            if (i > 0) {
                this.writer.write(", ");
            }
            this.emitNode(node.elements[i]);
        }
        this.writer.write("]");
        this.writer.write("</span>");
    };
    HtmlEmitter.prototype.emitParameter = function (node) {
        this.writer.write("<span class=\"parameter\">");
        this.emitIdentifier(node.name);
        this.writer.write("</span>");
    };
    HtmlEmitter.prototype.emitOneOfList = function (node) {
        this.writer.write("<span class=\"keyword\">one</span> <span class=\"keyword\">of</span>");
        this.writer.writeln();
        this.writer.write("<div class=\"one-of-list\">");
        for (var i = 0; i < node.terminals.length; ++i) {
            if (i > 0) {
                this.writer.write(" ");
            }
            this.writer.write("<span class=\"terminal\">");
            this.emitTextContent(node.terminals[i]);
            this.writer.write("</span>");
        }
        this.writer.write("</div>");
        this.emitTrailingHtmlTriviaOfNode(node);
        this.writer.writeln();
    };
    HtmlEmitter.prototype.emitRightHandSideList = function (node) {
        this.writer.write("<div class=\"rhs-list\">");
        this.writer.indent();
        for (var _i = 0, _a = node.elements; _i < _a.length; _i++) {
            var rhs = _a[_i];
            this.writer.writeln();
            this.writer.write("<div class=\"rhs-list-item\">");
            this.writer.indent();
            this.writer.writeln();
            this.emitNode(rhs);
            this.writer.dedent();
            this.writer.writeln();
            this.writer.write("</div>");
        }
        this.writer.dedent();
        this.writer.writeln();
        this.writer.write("</div>");
    };
    HtmlEmitter.prototype.emitRightHandSide = function (node) {
        var linkId = this.resolver.getRightHandSideLinkId(node, /*includePrefix*/ false);
        this.emitLinkAnchor(linkId);
        this.writer.write("<span class=\"rhs\">");
        this.emitNode(node.head);
        this.writer.write("</span>");
        this.emitTrailingHtmlTriviaOfNode(node);
        this.writer.writeln();
    };
    HtmlEmitter.prototype.emitSymbolSpan = function (node) {
        this.emitNode(node.symbol);
        if (node.next) {
            this.writer.write(" ");
            this.emitNode(node.next);
        }
    };
    HtmlEmitter.prototype.emitPlaceholder = function (node) {
        this.writer.write("<span class=\"placeholder\">@</span>");
    };
    HtmlEmitter.prototype.emitTerminal = function (node) {
        this.writer.write("<span class=\"terminal\">");
        this.emitTextContent(node);
        this.writer.write("</span>");
        if (node.questionToken) {
            this.writer.write("<span class=\"opt\">opt</span>");
        }
    };
    HtmlEmitter.prototype.emitNonterminal = function (node) {
        var linkId = this.resolver.getProductionLinkId(node.name);
        this.writer.write("<span class=\"nonterminal\">");
        this.emitNodeWithLink(node.name, linkId);
        this.writer.write("</span>");
        this.emitNode(node.argumentList);
        if (node.questionToken) {
            this.writer.write("<span class=\"opt\">opt</span>");
        }
    };
    HtmlEmitter.prototype.emitArgumentList = function (node) {
        this.writer.write("<span class=\"argument-list\">");
        this.writer.write("[");
        for (var i = 0; i < node.elements.length; ++i) {
            if (i > 0) {
                this.writer.write(", ");
            }
            this.emitNode(node.elements[i]);
        }
        this.writer.write("]");
        this.writer.write("</span>");
    };
    HtmlEmitter.prototype.emitArgument = function (node) {
        this.writer.write("<span class=\"argument\">");
        this.emitToken(node.operatorToken);
        this.emitNode(node.name);
        this.writer.write("</span>");
    };
    HtmlEmitter.prototype.emitUnicodeCharacterRange = function (node) {
        this.writer.write("<span class=\"unicode-character-range\">");
        this.emitUnicodeCharacterLiteral(node.left);
        this.writer.write(" <span class=\"keyword\">through</span> ");
        this.emitUnicodeCharacterLiteral(node.right);
        this.writer.write("</span>");
    };
    HtmlEmitter.prototype.emitUnicodeCharacterLiteral = function (node) {
        this.writer.write("<span class=\"unicode-character-literal\">");
        this.emitTextContent(node);
        this.writer.write("</span>");
        if (node.questionToken) {
            this.writer.write("<span class=\"opt\">opt</span>");
        }
    };
    HtmlEmitter.prototype.emitProse = function (node) {
        this.writer.write("<span class=\"prose\">");
        for (var _i = 0, _a = node.fragments; _i < _a.length; _i++) {
            var fragment = _a[_i];
            this.emitNode(fragment);
        }
        this.writer.write("</span>");
    };
    HtmlEmitter.prototype.emitEmptyAssertion = function (node) {
        this.writer.write("<span class=\"assertion\">[empty]</span>");
    };
    HtmlEmitter.prototype.emitSymbolSet = function (node) {
        this.writer.write("{");
        for (var i = 0; i < node.elements.length; ++i) {
            if (i > 0) {
                this.writer.write(",");
            }
            this.writer.write(" ");
            this.emitNode(node.elements[i]);
        }
        this.writer.write(" }");
    };
    HtmlEmitter.prototype.emitLookaheadAssertion = function (node) {
        this.writer.write("<span class=\"assertion\">[");
        switch (node.operatorToken.kind) {
            case tokens_1.SyntaxKind.ExclamationEqualsToken:
            case tokens_1.SyntaxKind.NotEqualToToken:
                this.writer.write("<span class=\"keyword\">lookahead</span> \u2260 ");
                break;
            case tokens_1.SyntaxKind.EqualsToken:
            case tokens_1.SyntaxKind.EqualsEqualsToken:
                this.writer.write("<span class=\"keyword\">lookahead</span> = ");
                break;
            case tokens_1.SyntaxKind.LessThanMinusToken:
            case tokens_1.SyntaxKind.ElementOfToken:
                this.writer.write("<span class=\"keyword\">lookahead</span> \u2208 ");
                break;
            case tokens_1.SyntaxKind.LessThanExclamationToken:
            case tokens_1.SyntaxKind.NotAnElementOfToken:
                this.writer.write("<span class=\"keyword\">lookahead</span> \u2209 ");
                break;
        }
        this.emitNode(node.lookahead);
        this.writer.write("]</span>");
    };
    HtmlEmitter.prototype.emitLexicalGoalAssertion = function (node) {
        this.writer.write("<span class=\"assertion\">[<span class=\"keyword\">lexical</span> </span>goal</span>");
        var linkId = this.resolver.getProductionLinkId(node.symbol);
        this.emitNodeWithLink(node.symbol, linkId);
        this.writer.write("</span>");
    };
    HtmlEmitter.prototype.emitNoSymbolHereAssertion = function (node) {
        this.writer.write("<span class=\"assertion\">[<span class=\"keyword\">no</span> ");
        if (node.symbols) {
            for (var i = 0; i < node.symbols.length; ++i) {
                if (i > 0) {
                    this.writer.write(" <span class=\"keyword\">or</span> ");
                }
                this.emitNode(node.symbols[i]);
            }
        }
        this.writer.write(" <span class=\"keyword\">here</span>]</assertion>");
    };
    HtmlEmitter.prototype.emitParameterValueAssertion = function (node) {
        this.writer.write("<span class=\"assertion\">[");
        this.emitToken(node.operatorToken);
        this.writer.write("<span class=\"parameter\">");
        this.emitNode(node.name);
        this.writer.write("</span>");
        this.writer.write("]</span>");
    };
    HtmlEmitter.prototype.emitProseAssertion = function (node) {
        this.writer.write("<span class=\"assertion\">[");
        for (var _i = 0, _a = node.fragments; _i < _a.length; _i++) {
            var fragment = _a[_i];
            this.emitNode(fragment);
        }
        this.writer.write("]</span>");
    };
    HtmlEmitter.prototype.emitButNotSymbol = function (node) {
        this.emitNode(node.left);
        this.writer.write(" <span class=\"keyword\">but</span> <span class=\"keyword\">not</span> ");
        this.emitNode(node.right);
    };
    HtmlEmitter.prototype.emitOneOfSymbol = function (node) {
        this.writer.write("<span class=\"keyword\">one</span> <span class=\"keyword\">of</span> ");
        if (node.symbols) {
            for (var i = 0; i < node.symbols.length; ++i) {
                if (i > 0) {
                    this.writer.write(" <span class=\"keyword\">or</span> ");
                }
                this.emitNode(node.symbols[i]);
            }
        }
    };
    HtmlEmitter.prototype.emitTextContent = function (node) {
        if (node) {
            var text = node.text;
            this.writer.write(this.encode(text));
        }
    };
    HtmlEmitter.prototype.emitLinkAnchor = function (linkId) {
        if (linkId && this.options.emitLinks) {
            this.writer.write("<a name=\"" + linkId + "\"></a>");
        }
    };
    HtmlEmitter.prototype.emitNodeWithLink = function (node, linkId) {
        if (linkId && this.options.emitLinks) {
            this.writer.write("<a href=\"#" + linkId + "\">");
            this.emitNode(node);
            this.writer.write("</a>");
        }
        else {
            this.emitNode(node);
        }
    };
    return HtmlEmitter;
}(emitter_1.Emitter));
exports.HtmlEmitter = HtmlEmitter;
